### ISSUE TITLE
Bad UI performance on "Add Dag Run" (/dagrun/add) with many task instances

### ISSUE THREAD

**Issue by Gerrit-K** [apache/airflow#37026]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.8.0

### What happened?

When accessing the "Add Dag Run" dialog (`<airflow-URL>/dagrun/add`, see screenshot below) we're facing bad UI performance (>5s response time). Upon further inspection, we realized that there were two select fields in the response that were pre-populated with _a lot_ of options: Task Instances and Creating Job. Both of them contain around 20k values in our case, which (1) heavily increases the response time and (2) makes the select box almost unusable in the browser once the page eventually renders.

![image](https://github.com/apache/airflow/assets/8766565/cb0b9ef7-085d-4058-8ec8-c7a9e6bba544)

![image](https://github.com/apache/airflow/assets/8766565/fb6cec8b-abcf-458e-9858-7874257e3351)

![image](https://github.com/apache/airflow/assets/8766565/2af04746-a70b-4364-a2e9-a1ffb81dc3b8)

### How to reproduce

Create a larger number of tasks (e.g. by running a long backfill of a multi-task DAG). Then access the mentioned view.

### Operating System

debian 12 (bookworm)

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

Official helm chart with official airflow image (2.8.0)

### Anything else?

Although 20k+ task instances sounds like a lot, I think this can easily happen for a production environment. However, I'm just starting to dive into Airflow, so if it is indeed wrong to have that many task instances (e.g. if they should be cleaned up) please let me know.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Yes. you can clean stuff with `airflow db clean` command (see the docs).. Not sure if the current behaviour can be easily improved. One way to do it is to replace the drop-down with search box if the number of distinct entities is reaching a threshold. Marking as good first issue, might be a good issue for someone who does web/ui stuff

**Comment by Gerrit-K** [comment]
> Yes. you can clean stuff with `airflow db clean` command (see the docs)

Ah, I see. Would I still be able to run backfills for dates that have been cleaned up in this case?

> One way to do it is to replace the drop-down with search box if the number of distinct entities is reaching a threshold.

Sounds like a good idea. I don't know what capabilities the Airflow UI framework has, but maybe there is already a combobox component that could be used here.

**Comment by Gerrit-K** [comment]
I just tested with Airflow v2.10.4 and can't find the blue plus button anymore and accessing `/dagrun/add` manually resulted in a 404, so I assume this view has been removed. I didn't check the release notes yet, but it's also not of importance for us anymore.
